tower-balance/src/pool.rs

```diff
@@ -14,311 +14,117 @@
-    type Error = <Balance<PoolDiscoverer<MS, Target, Req>, Req> as Service<Req>>::Error;
-    type Future = <Balance<PoolDiscoverer<MS, Target, Req>, Req> as Service<Req>>::Future;
+    type Response = <MS::Service as Service<Req>>::Response;
+    type Error = error::Error;
+    type Future = <MS::Service as Service<Req>>::Future;
 
     fn poll_ready(&mut self) -> Poll<(), Self::Error> {
-        if let Async::Ready(()) = self.balance.poll_ready()? {
-            // services was ready -- there are enough services
-            // update ewma with a 0 sample
-            self.ewma = (1.0 - self.options.alpha) * self.ewma;
-
-            let discover = self.balance.discover_mut();
-            if self.ewma < self.options.low {
-                discover.load = Level::Low;
-
-                if discover.services > 1 {
-                    // reset EWMA so we don't immediately try to remove another service
-                    self.ewma = self.options.init;
-                }
-            } else {
-                discover.load = Level::Normal;
+        let adding = match self.pending_service.take() {
+            None => false,
+            Some(fut) => {
+                self.poll_pending_service(fut)?;
+                true
             }
+        };
 
+        // Drive pending services, including any just-added services, to ready.
+        // If there are ready services, we're ready to dispatch a request
+        // without doing any more work.
+        self.cache.poll_unready();
+        if self.cache.ready_len() > 0 {
             return Ok(Async::Ready(()));
         }
 
-        let discover = self.balance.discover_mut();
-        if discover.making.is_none() {
-            // no services are ready -- we're overloaded
-            // update ewma with a 1 sample
-            self.ewma = self.options.alpha + (1.0 - self.options.alpha) * self.ewma;
+        if adding {
+            // A service is already pending or was just added to the cache, so
+            // defer until an unready service notifies availability.
+            return Ok(Async::NotReady);
+        }
+        debug_assert!(self.pending_service.is_none());
 
-            if self.ewma > self.options.high {
-                discover.load = Level::High;
+        if self.cache.len() == self.max_services {
+            // The cache is at capacity so we're stuck waiting until new an
+            // existing service is available.
+            return Ok(Async::NotReady);
+        }
 
-                // don't reset the EWMA -- in theory, poll_ready should now start returning
-                // `Ready`, so we won't try to launch another service immediately.
-                // we clamp it to high though in case the # of services is limited.
-                self.ewma = self.options.high;
-            } else {
-                discover.load = Level::Normal;
-            }
+        let fut = self.make_service.make_service(());
```

Comment:
This will create new services _very_ aggressively. What's the reasoning behind not maintaining a load estimate?

Reply:
That depends entirely on how inner services advertise readiness. If, for instance, an inner service uses a buffer, a new service would only ever be created when the buffer is at capacity.

This is still an experiment, obviously, but i'm curious if we avoid having `pool` dependent on a specific load metric implementation.  (Note, for instance, that the pool ewma doesn't really take into account time-to-first-byte or errors or other data that may be measured to gauge health/load).